File: example-projects/csharp-basic/src/Utils/PriceCalculator.cs
namespace CSharpBasic.Utils;

public static class PriceCalculator
{
    public static decimal ApplyDiscount(decimal price, decimal discountPercent)
    {
        if (discountPercent is < 0 or > 100)
            throw new ArgumentOutOfRangeException(nameof(discountPercent), "Must be between 0 and 100.");

        return price * (1 - discountPercent / 100m);
    }

    public static decimal RoundToNearest(decimal price, decimal step = 0.05m)
        => Math.Round(price / step) * step;
}

File: example-projects/rust-basic/src/utils/price_calculator.rs
/// Applies a percentage discount to a price.
pub fn apply_discount(price: f64, discount_percent: f64) -> f64 {
    assert!((0.0..=100.0).contains(&discount_percent), "Discount must be between 0 and 100");
    price * (1.0 - discount_percent / 100.0)
}

/// Rounds a price to the nearest step (default 0.05).
pub fn round_to_nearest(price: f64, step: f64) -> f64 {
    (price / step).round() * step
}
